File: cpp/src/lexer/token.cpp
#include <ostream>

#include "token.hpp"


std::string std::to_string(TokenType tokenType)
{
	switch (tokenType) {
		case TokenType::Illegal:        return "(illegal)";
		case TokenType::Eof:            return "(eof)";
		case TokenType::Identifier:     return "(identifier)";

		case TokenType::Integer:        return "(integer)";
		case TokenType::Comma:          return ",";
		case TokenType::Semicolon:      return ";";
		case TokenType::Lparen:         return "(";
		case TokenType::Rparen:         return ")";
		case TokenType::Lsquirly:       return "{";
		case TokenType::Rsquirly:       return "}";
		case TokenType::Lbracket:       return "[";
		case TokenType::Rbracket:       return "]";

		// OpTokenType::erator:         return "";
		case TokenType::Assign:         return "=";
		case TokenType::Plus:           return "+";
		case TokenType::Minus:          return "-";
		case TokenType::Bang:           return "!";
		case TokenType::Asterisk:       return "*";
		case TokenType::Slash:          return "/";
		case TokenType::Percent:        return "%";
		case TokenType::Tilde:          return "~";
		case TokenType::BitAnd:         return "&";
		case TokenType::BitOr:          return "|";
		case TokenType::BitEor:         return "^";
		case TokenType::And:            return "&&";
		case TokenType::Or:             return "||";

		case TokenType::Lt:             return "<";
		case TokenType::Gt:             return ">";
		case TokenType::Le:             return "<=";
		case TokenType::Ge:             return ">=";

		// KeTokenType::yword:          return "";
		case TokenType::Function:       return "fn";
		case TokenType::Let:            return "let";
		case TokenType::True:           return "true";
		case TokenType::False:          return "false";
		case TokenType::If:             return "if";
		case TokenType::Else:           return "else";
		case TokenType::Return:         return "return";

		case TokenType::Eq:             return "==";
		case TokenType::Not_eq:         return "!=";

		case TokenType::String:         return "(String)";
	}
	return std::string{"(unhandled "} + std::to_string((int) tokenType) + ")";
}


std::ostream& operator<<(std::ostream& os, const TokenType& tokenType)
{
	os << std::to_string(tokenType);
	return os;
}

std::ostream& operator<<(std::ostream& os, const Token& token)
{
	switch (token.type) {
		case TokenType::Identifier: os << token.literal; break;
		case TokenType::Integer:    os << token.literal; break;
		case TokenType::String:     os << "\"" << token.literal << "\""; break;
		default:                    os << token.type;    break;
	}
	return os;
}

File: dart/lib/src/lexer/token.dart
enum TokenType {
  illegal,
  eof,

  equal,
  assign,
  notEqual,
  bang,
  plus,
  minus,
  slash,
  asterisk,
  lte,
  gte,
  lt,
  gt,
  comma,
  semicolon,

  lParen,
  rParen,
  lSquirly,
  rSquirly,

  ident,
  int,

  function,
  let,
  true_,
  false_,
  if_,
  else_,
  return_,
  ;

  const TokenType();
}

class Token {
  final TokenType type;
  final String literal;

  const Token(this.type, this.literal);

  const Token.illegal(this.literal) : type = TokenType.illegal;
  const Token.eof() : this(TokenType.eof, "");

  const Token.equal() : this(TokenType.equal, "==");
  const Token.assign() : this(TokenType.assign, "=");
  const Token.notEqual() : this(TokenType.notEqual, "!=");
  const Token.bang() : this(TokenType.bang, "!");
  const Token.plus() : this(TokenType.plus, "+");
  const Token.minus() : this(TokenType.minus, "-");
  const Token.slash() : this(TokenType.slash, "/");
  const Token.asterisk() : this(TokenType.asterisk, "*");
  const Token.lte() : this(TokenType.lte, "<=");
  const Token.gte() : this(TokenType.gte, ">=");
  const Token.lt() : this(TokenType.lt, "<");
  const Token.gt() : this(TokenType.gt, ">");
  const Token.comma() : this(TokenType.comma, ",");
  const Token.semicolon() : this(TokenType.semicolon, ";");

  const Token.lParen() : this(TokenType.lParen, "(");
  const Token.rParen() : this(TokenType.rParen, ")");
  const Token.lSquirly() : this(TokenType.lSquirly, "{");
  const Token.rSquirly() : this(TokenType.rSquirly, "}");

  const Token.ident(this.literal) : type = TokenType.ident;
  const Token.int(this.literal) : type = TokenType.int;
  const Token.fn() : this(TokenType.function, "fn");
  const Token.let() : this(TokenType.let, "let");
  const Token.true_() : this(TokenType.true_, "true");
  const Token.false_() : this(TokenType.false_, "false");
  const Token.if_() : this(TokenType.if_, "if");
  const Token.else_() : this(TokenType.else_, "else");
  const Token.return_() : this(TokenType.return_, "return");

  @override
  String toString() => "Token.${type.name}('$literal')";

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Token && type == other.type && literal == other.literal;

  @override
  int get hashCode => type.hashCode ^ literal.hashCode;
}
